### PR TITLE
fb: cfb: Add basic drawing APIs 

### PR DESCRIPTION
**Pull request by soburi**
Add basic graphics(point, line rectangle) drawing APIs.
Also add shell commands.

### CODE REVIEW

File: subsys/fb/cfb.c [hunk 1]
```diff
@@ -299,6 +299,38 @@ static inline void draw_point(struct char_framebuffer *fb, int16_t x, int16_t y)
 	draw_horizontal_line(fb, x, x + 1, y);
 }
 
+static void draw_line(struct char_framebuffer *fb, int16_t x0, int16_t y0, int16_t x1, int16_t y1)
+{
+	/* Bresenham's line algorithm */
+
```
[the remark this answers is not in the document]
Comment (soburi) [reply subsys/fb/cfb.c]: Fixed.

File: subsys/fb/cfb.c [hunk 2]
```diff
@@ -299,6 +299,38 @@ static inline void draw_point(struct char_framebuffer *fb, int16_t x, int16_t y)
 	draw_horizontal_line(fb, x, x + 1, y);
 }
 
+static void draw_line(struct char_framebuffer *fb, int16_t x0, int16_t y0, int16_t x1, int16_t y1)
```
[the remark this answers is not in the document]
Comment (soburi) [reply subsys/fb/cfb.c:302]: For some needs, I create draw_vertical_line and draw_horizontal_line. draw_vertical_line can be shared with draw_point, draw_horizontal_line will be convinicence for filling.
It's a lightweight implementation, so it should be better for rectangles that don't draw slant lines.

**Comment by soburi** [comment]
@jfischer-no 

> @soburi Looks good at first glance. It would be nice to get rid of the framebuffer as such and write directly to the controller's memory using an intermediate buffer with a minimum size, maybe of what the controller supports for a memory write access. Any thoughts? It is not a blocker for this PR.

It is a too big task for treating in this PR.
I'll consider it in the near future.
I think that many drivers need to implement a `read` operation to work this design effectively.
If not, A draw operation loses what has already been drawn. It need 'read-modify-write' operation.

File: subsys/fb/cfb.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply subsys/fb/cfb.c:302]: > It's a lightweight implementation, so it should be better for rectangles that don't draw slant lines.

A lightweight implementation should not contain redundant code/functions. Also, at some point we have to get rid of the large buffer inside `struct char_framebuffer *fb`, the less primitive we have the easier it becomes.

[the remark this answers is not in the document]
Comment (soburi) [reply subsys/fb/cfb.c:207]: I got it.
Change to don't create draw_horizontal_line and draw_vertical_line. Use draw_line instead of these.

**Comment by danieldegrasse** [comment]
@soburi Just a question on the goals of this PR- is this primarily intended to offer helper functions for drawing monochrome buffers? I ask because I'm investigating how we can better support graphics acceleration in Zephyr, and want to understand if there is any overlap between the goals of this PR and the work I'm currently doing.

**Comment by soburi** [comment]
@danieldegrasse 

> @soburi Just a question on the goals of this PR- is this primarily intended to offer helper functions for drawing monochrome buffers? I ask because I'm investigating how we can better support graphics acceleration in Zephyr, and want to understand if there is any overlap between the goals of this PR and the work I'm currently doing.

Yes, it is.
I want to add graphics API for memory-constrained devices.
In general, Zephyr can use legal. But I think the implementation is not optimum for monochrome display from the viewpoint of memory consumption.

**Comment by danieldegrasse** [comment]
> I want to add graphics API for memory-constrained devices.

If you're going to be at EOSS this year, you should come to my talk: https://sched.co/1LxCs. I'm also looking at creating an API for graphics in Zephyr, and would love to discuss ideas with you! If you're not attending, I would still appreciate your comments on the eventual RFC. Though I am primarily focused on hardware acceleration, I think that any API for this functionality should capture both software and hardware rendering use cases.

**Comment by soburi** [comment]
@danieldegrasse 

Thank you for your comment.

I have yet to plan to attend EOSS now.
Czech is a bit far from Japan... :-(
I'll check the session If the video is published. (LF event usually publishes the recording, I expect it.)

- - - 

In my current opinion...

Monochrome displays are not a bit friendly for hardware acceleration.
It is not an 'a byte per a dot'. So it usually needs special consideration.

I think the CFB subsystem may be suitable for systems that can't use LVGL.

I think LVGL is excellent for standard graphics subsystem.
But, some systems can't use LVGL (For example, in case that system has too few memory for allocate framebufer.).

I think it is good the CFB subsystem keep for systems that cannot use LVGL.
In this case, It may be possible to design it as a "unified API" with hardware accelaration, but I'm also worried that the implementation will be completely different for monochrome and color.

**Comment by danieldegrasse** [comment]
> I think it is good the CFB subsystem keep for systems that cannot use LVGL.
In this case, It may be possible to design it as a "unified API" with hardware acceleration, but I'm also worried that the implementation will be completely different for monochrome and color.

I agree here- I think systems that render to color displays will need more complicated API then monochrome displays use. This is especially true because implementations for color graphics may be more concerned about performance than code size. For systems with limited memory, I think the extensions you have proposed here make sense.

**Review by jfischer-no — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**